### PR TITLE
Fast standalone symbolize for unwinding

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123966

We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

Folly Symbolize or LLVMSymbolize are fast but it requires PyTorch take a dependency on those libraries to do this, and given the number of environments we run stuff in, we end up hitting cases where we fallback to slow addr2line behavior.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
no external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

I've tested this on some memory profiling runs using all combinations of {gcc, clang} x {dwarf4, dwarf5} and it seems to do a good job at getting line numbers and function names right. It is also careful to route all reads of library data through the `CheckedLexer` object, which ensure it is not reading out of bounds of the section. Errors are routed through UnwindError so that those exceptions get caught and we produce a ?? frame rather than crash. I also added a fuzz test which gives all our symbolizer options random addresses in the process to make sure they do not crash.

Differential Revision: [D56828968](https://our.internmc.facebook.com/intern/diff/D56828968)

### CODE REVIEW

File: test/profiler/test_profiler.py [hunk 1]
```diff
@@ -3561,6 +3566,70 @@ def test_profiler_pattern_matcher_json_report(self):
...
+                        section_name = mm[name_start : name_start + 6]
+                        if section_name != b".text\0":
+                            continue
+                        (section_offset,) = unpack(
+                            "Q", section_headers_start + i * section_header_size + 24
+                        )
+                        (section_size,) = unpack(
+                            "Q", section_headers_start + i * section_header_size + 32
+                        )
+                        start = int(filename.split("-")[0], 16) + section_offset
```
Comment (wconstab) [inline test/profiler/test_profiler.py:3612]: what is the filename representing- does each mapped library have a filename representing its start address in memory?

File: test/profiler/test_profiler.py [hunk 2]
```diff
@@ -3561,6 +3566,70 @@ def test_profiler_pattern_matcher_json_report(self):
...
+                        text_sections.append((start, section_size))
+                        break
+                    mm.close()
+            return text_sections
+
+        r = random.Random()
+        r.seed(1)
+        text_sections = get_text_sections()
+        addrs = []
+        for i in range(200):
```
Comment (wconstab) [inline test/profiler/test_profiler.py:3622]: ok so basically, take any random byte of code and ask for its unwind, make sure this doesn't crash.  seems reasonable

File: torch/csrc/profiler/unwind/dwarf_symbolize_enums.h [hunk 3]
```diff
@@ -0,0 +1,181 @@
+#pragma once
+#include <torch/csrc/profiler/unwind/unwind_error.h>
+#include <cstdint>
+#include <optional>
+
+enum {
+  DW_TAG_subprogram = 0x2e,
```
Comment (wconstab) [inline torch/csrc/profiler/unwind/dwarf_symbolize_enums.h:7]: isn't there some dwarf header we can include that defines all these things? i guess we don't have access to it though?

File: torch/csrc/utils/cpp_stacktraces.cpp [hunk 4]
```diff
@@ -47,9 +47,31 @@ bool get_cpp_stacktraces_enabled() {
   return enabled;
 }
 
-bool get_disable_addr2line() {
-  static bool disabled = compute_disable_addr2line();
-  return disabled;
+static torch::unwind::Mode compute_symbolize_mode() {
+  auto envar_c = std::getenv("TORCH_SYMBOLIZE_MODE");
```
Comment (wconstab) [inline torch/csrc/utils/cpp_stacktraces.cpp:51]: we already ensure this env is set to something? or where is the default provided

File: torch/csrc/utils/cpp_stacktraces.cpp [hunk 5]
```diff
@@ -47,9 +47,31 @@ bool get_cpp_stacktraces_enabled() {
...
+    } else if (envar == "fast") {
+      return unwind::Mode::fast;
+    } else {
+      TORCH_CHECK(
+          false,
+          "expected {dladdr, addr2line, fast} for TORCH_SYMBOLIZE_MODE, got ",
+          envar);
+    }
+  } else {
+    return compute_disable_addr2line() ? unwind::Mode::dladdr
```
Comment (wconstab) [inline torch/csrc/utils/cpp_stacktraces.cpp:67]: i assume this is the status quo today, we default to dladdr only if we set the env for addr2line disablement

**Comment by ezyang** [comment]
Not sure how you want to run review here, an actual careful review will take some time since I don't know how DWARF works. Shall we just dropship the code in?

File: test/profiler/test_profiler.py (hunk 1 above)
Comment (zdevito) [reply test/profiler/test_profiler.py:3612]: A file looks like this:
`lr-------- 1 zdevito users 64 Apr 22 09:11 7fcea5800000-7fcea5828000 -> /usr/lib64/libc.so.6`

File: torch/csrc/profiler/unwind/dwarf_symbolize_enums.h (hunk 3 above)
Comment (zdevito) [reply torch/csrc/profiler/unwind/dwarf_symbolize_enums.h:7]: It doesn't come by default. However the numbers themselves are defined by the standard, so they can't change anyway.

File: torch/csrc/utils/cpp_stacktraces.cpp (hunk 5 above)
Comment (zdevito) [reply torch/csrc/utils/cpp_stacktraces.cpp:67]: yeah, and once this lands we can delete DISABLE_ADDR2LINE in favor of just setting the mode. Hopefully we can enable the fast version by default and get rid of the flags.

**Comment by zdevito** [comment]
For review, I think it is fine if the review just makes sure this doesn't break anything when disabled, and we can land it an experiment with it on some real runs by setting environment flags. No need to go deeply into DWARF. If you have time, a quick check looking for "this can't crash if the dwarf information is bogus" would be good. I tried to put all accesses through the `CheckedLexer` which can't read out of bounds of a segment.

File: torch/csrc/profiler/unwind/unwind_error.h [hunk 6]
```diff
@@ -1,6 +1,31 @@
...
+  } while (0)
+
+// #define LOG_INFO(...) fmt::print(__VA_ARGS__)
+#define LOG_INFO(...)
+
+// #define PRINT_INST(...) LOG_INFO(__VA_ARGS__)
+#define PRINT_INST(...)
+
+// #define PRINT_LINE_TABLE(...) LOG_INFO(__VA_ARGS__)
+#define PRINT_LINE_TABLE(...)
```
Comment (ezyang) [inline torch/csrc/profiler/unwind/unwind_error.h:27]: would be nice if these could go into our glog logging permanently as debug but whatever lol

File: torch/csrc/profiler/python/init.cpp [hunk 7]
```diff
@@ -606,6 +603,33 @@ void initPythonBindings(PyObject* module) {
...
+      "symbolize_addresses",
+      [](const std::vector<uint64_t>& frames, const std::string& mode_s) {
+        std::vector<std::tuple<std::string, int64_t, std::string>> frames_out;
+        torch::unwind::Mode mode = torch::unwind::Mode::addr2line;
+        if (mode_s == "fast") {
+          mode = torch::unwind::Mode::fast;
+        } else if (mode_s == "addr2line") {
+          mode = torch::unwind::Mode::addr2line;
+        } else if (mode_s == "dladdr") {
+          mode = torch::unwind::Mode::dladdr;
```
Comment (ezyang) [inline torch/csrc/profiler/python/init.cpp:617]: a bit dupey with the envvar parsing code

**Review by ezyang — APPROVED** [review]
Integration code looks fine. I didn't review contents of unwind/ dir

**Comment by zdevito** [comment]
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56828968).

**Review by aaronenyeshi — APPROVED** [review]

**Pull request closed without merging**